flex_model/core/hook_function.py

```diff
@@ -296,96 +291,98 @@ def _dispatch_hook_function(
-        outputs = self._template_for_input_output_editing(module, outputs)
+        outputs = self._template_handle_layer_outputs(module, outputs)
         return outputs
 
-    def _handle_backward(
+    def _full_backward_hook_impl(
         self,
         module: nn.Module,
         grad_inputs: Union[LayerOutputs, Tensor],
         _grad_outputs: Union[LayerOutputs, Tensor],
     ) -> Union[LayerOutputs, Tensor]:
         """Runs a hook function for editing backward module input gradients."""
-        outputs = self._template_for_input_output_editing(module, grad_inputs)
+        outputs = self._template_handle_layer_outputs(module, grad_inputs)
+        return outputs
+
+    def _tensor_hook_impl(self, grad: Tensor,) -> Tensor:
+        """Runs a hook function for editing tensor gradients."""
+        # No module since this is tensor-level.
+        outputs = self._template_handle_tensor(None, grad)
         return outputs
 
-    def _handle_pre_forward(
+    def _forward_pre_hook_impl(
         self, module: nn.Module, args: Union[LayerOutputs, Tensor],
     ) -> Union[LayerOutputs, Tensor]:
         """Runs a hook function for editing forward module inputs."""
         # Same procedure as `_handle_forward`, just that we operate on args.
-        outputs = self._template_for_input_output_editing(module, args)
+        outputs = self._template_layer_outputs(module, args)
         return outputs
 
-    def _handle_pre_backward(
+    def _full_backward_pre_hook_impl(
         self, module: nn.Module, grad_outputs: Union[LayerOutputs, Tensor],
     ) -> Union[LayerOutputs, Tensor]:
         """Runs a hook function for editing backward module output gradients."""
-        outputs = self._template_for_input_output_editing(module, grad_outputs)
-        return outputs
-
-    def _handle_tensor_backward(self, grad: Tensor,) -> Tensor:
-        """Runs a hook function for editing tensor gradients."""
-        # No module since this is tensor-level.
-        outputs = self._template_for_point_tensor_editing(None, grad)
+        outputs = self._template_layer_outputs(module, grad_outputs)
         return outputs
 
-    def _template_for_point_tensor_editing(
-        self, module: nn.Module, tensor: Tensor,
-    ) -> Tensor:
+    def _template_handle_tensor(self, module: nn.Module, tensor: Tensor,) -> Tensor:
         """Template function for editing a sharded activation tensor.
 
         This function is used alone in cases where hook functions operate
         directly on a tensor, and not an entire module.
         """
         start_shape = tensor.shape
 
-        # Poplate the collection and dispersion functions.
-        if self._collect is None or self._disperse is None:
-            self._parse_tensor(tensor)
+        # Concretize functions.
+        if (
+            self._collect is None
+            or self._disperse is None
+            or self._edit is None
+            or self._offload is None
+        ):
```

Comment:
There should be a check to make sure these are either 1. All `None` or 2. All `callable`. Any mixed states could lead to silent failures and should be explicitly invalid.